guide/src/module.md

```diff
@@ -48,53 +53,110 @@ To import the module, either:
 
 ## Documentation
 
-The [Rust doc comments](https://doc.rust-lang.org/stable/book/first-edition/comments.html) of the module
+The [Rust doc comments](https://doc.rust-lang.org/stable/book/ch03-04-comments.html) of the module
 initialization function will be applied automatically as the Python docstring of your module.
 
+For example, building off of the above code, this will print `This module is implemented in Rust.`:
+
 ```python
-import rust2py
+import my_extension
 
-print(rust2py.__doc__)
+print(my_extension.__doc__)
 ```
 
-Which means that the above Python code will print `This module is implemented in Rust.`.
+## Organizing your module registration code
```

Comment:
I would probably avoid this indirection and just add the classes and functions directly:

```rust
#[pymodule]
fn my_extension(py: Python, m: &PyModule) -> PyResult<()> {
    m.add_class::<dirutil::SomeClass>()?;
    m.add_function(wrap_pyfunction!(osutil::determine_current_os, m)?)?;
    Ok(())
}
```

I guess this makes more sense if you have a big hierarchy of modules with register functions that each import other registers? But at that point you should probably just organize them in submodules.

Reply:
I remember now why I originally came up with this pattern of passing `&PyModule` around:

```rust
error: no rules expected the token `::`
  --> src/lib.rs:591:43
   |
14 |     m.add_function(wrap_pyfunction!(osutil::determine_current_os, m)?)?;
   |                                           ^^ no rules expected this token in macro call
```

Is that expected? Any known workarounds other than what I'm suggesting?

Reply:
Hm my next idea didn't work either to leverage `use`:

```rust
// src/lib.rs
use pyo3::prelude::*;
use osutil::determine_current_os; 

#[pymodule]
fn my_extension(py: Python, m: &PyModule) -> PyResult<()> {
    m.add_class::<dirutil::SomeClass>()?;
    m.add_function(wrap_pyfunction!(determine_current_os, m)?)?;
    Ok(())
}

// src/dirutil.rs
# mod dirutil { 
use pyo3::prelude::*;

#[pyclass]
pub(crate) struct SomeClass {
    x: usize,
}
# }

// src/osutil.rs
# mod osutil { 
use pyo3::prelude::*;

#[pyfunction]
pub(crate) fn determine_current_os() -> String {
    "linux".to_owned()
}  
# }
```

```
❯ cargo test --doc -- doc_test::guide_module_md
warning: profiles for the non root package will be ignored, specify profiles at the workspace root:
package:   /Users/eric/code/pyo3/examples/pyo3-pytests/Cargo.toml
workspace: /Users/eric/code/pyo3/Cargo.toml
    Finished test [unoptimized + debuginfo] target(s) in 0.10s
   Doc-tests pyo3

running 5 tests
test src/lib.rs - doc_test::guide_module_md (line 620) ... ignored
test src/lib.rs - doc_test::guide_module_md (line 578) ... FAILED
test src/lib.rs - doc_test::guide_module_md (line 535) ... ok
test src/lib.rs - doc_test::guide_module_md (line 513) ... ok
test src/lib.rs - doc_test::guide_module_md (line 670) ... ok

failures:

---- src/lib.rs - doc_test::guide_module_md (line 578) stdout ----
error[E0425]: cannot find value `__pyo3_get_function_determine_current_os` in this scope
  --> src/lib.rs:586:20
   |
11 |     m.add_function(wrap_pyfunction!(determine_current_os, m)?)?;
   |                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ not found in this scope
   |
   = note: this error originates in a macro (in Nightly builds, run with -Z macro-backtrace for more info)

error: aborting due to previous error
```

Reply:
Yeah, you're running into https://github.com/PyO3/pyo3/issues/1709

For now you can work around this by writing `use osutil::*;`

(As stated in the issue above, I hope we can fix it in a nicer way at the end of the year when we do a dependency update release.)